### PR TITLE
GOES-16/17 Converter - Enhanced MetaData

### LINKED ISSUE #615: Addition of New MetaData Variables to the GOES-16/17 IODAv2 Converter
## Description

New MetaData Variables must be added to the GOES-16/17 IODAv2 converter for utilization in EWOK and JEDI-GDAS for the NOAA Data Lake project.

## Acceptance Criteria (Definition of Done)

This issue is complete when successful execution of a JEDI-GDAS HofX3D EWOK experiment is completed.

### LINKED ISSUE #618: Addition of Data Subsampling to GOES-16/17 IODAv2 Converter
## Description

Due to long computational wall times in JEDI-GDAS EWOK experiments created by ~29M `nlocs` per channel in resulting GOES-16/17 files for the abi_g16 and abi_g17 observation types, subsampling of the resulting data files must be done to reduce the number of observations per scan. The current resolution is the 2km and is the native resolution for GOES-16/17 channels other than 1, 3, 5 (1km) and channel 2 (0.5 km). All data converted to IODAv2 using this converter are at 2km with channels 1, 2, 3, and 5 subsampled to the standard 2km resolution. 

## Requirements

The converter must utilize an optional keyword argument that will be used to subsample results accurately. 

## Acceptance Criteria (Definition of Done)

This issue is completed when the GOES-16/17 IODAv2 converter can subsample input datasets globally to even number resolutions (e.g., 4km, 8km, 16km)

### LINKED ISSUE #619: Removal of Nonexistent Lat/Lon and Data Values in the GOES16/17 IODAv2 Converter
## Description
The GOES16/17 IODAv2 Converter should not write any data values associated with nonexistent Lat/Lon coordinates. 

## Acceptance Criteria (Definition of Done)
This issue is complete when all nonexistent Lat/Lon and associated data values are removed from the GOES16/17 IODAv2 data files.

### LINKED ISSUE #624: Addition of Boolean Flag to Indicate Whether to Include Reflectance Factor in GOES-16/17 Conversion
## Description

This issue addresses the need to possibly exclude reflectance factor from GOES-16/17 IODAv2 conversion. 

## Acceptance Criteria (Definition of Done)

The GOES-16/17 IODAv2 converter accepts a boolean parameter to include or exclude reflectance factor generation.

### LINKED ISSUE #625: Verify Sensor and Solar Angle Formulas Used in the GOES-16/17 Converter
## Description

This issue addresses the need to verify the formulas used in the GOES-16/17 IODA converter for sensor and solar angles and whether change to the code are required. 

## Acceptance Criteria (Definition of Done)

This issue is complete when the formulas have been verified as the correct ones used for sensor angles.

### PR DESCRIPTION
**Pull request by ericlingerfelt**
## Description

This pull request contains additions to the GOES-16/17 IODAv2 converter and adds the following MetaData variables. 

/MetaData/scan_position
/MetaData/sensor_azimuth_angle
/MetaData/sensor_view_angle
/MetaData/sensor_zenith_angle
/MetaData/solar_azimuth_angle
/MetaData/solar_zenith_angle

This pull request also adds a `reference` directory which contains all documents used to derive formulations and other information used in the converter.

This pull request updates the converter to use a provided resolution for sub-sampling the data from 2km to 256km for smaller file sizes and observation profiles. The new GoesUtil class consolidates the sub-sampling and yaw_flip_flag filter. The GoesLatLon class has been modified to correctly account for longitude coordinates less than -180 degrees. The converter also now removes all nonexistent Lat/Lon values and any data values associated with the bad coordinates. 

### Issue(s) addressed

Link the issues to be closed with this PR
- fixes #615 
- fixes #618 
- fixes #619 
- fixes #624 
- fixes #625

### CODE REVIEW

**Review by srherbener — COMMENTED** [review]
Thanks for these updates. In general I think the code looks good. I have the following comments:


----------
* The CI test failures are due to a coding norms error. Click on one of the CDash links, then scroll to the bottom of the webpage and click on the link to view test results. The coding norms test should be highlighted (red) for the test failure and you can click on that to see the coding norms error.
---------
* It would be good to add a short unit test at some point. I'm okay with doing that in a subsequent PR.
* The code has a dependency on the solo repository, but it looks like all that is being used out of the solo repo are a couple datetime utilities (Date, JediDate). Would it be possible to use the python datetime module. The reason I'm asking is that it seems a bit heavy-weight to add another dependency to ioda just for a couple datetime utilities. But, if there is something about the solo utilities that is essential (eg, performance) then it's fine to use them.
* If it's okay to switch to the python datetime utilities, then adding a test will be easy and can be done relatively soon.
* If we need to stay with the solo datetime utilities, then we will need to introduce solo into ioda-bundle and include it when we build ioda-converters. Then we can add the test.
----------
* I'm not sure if its good to keep a copy of the references in with the repo.
    * The advantages I see of keeping a copy of the references:
        * They are readily available for people to read
    * The disadvantages I see are: 
        * As you make code changes based on updates with procedures at the data provider, you need to update the references so this introduces sync'ing issues and maintenance tasks
        *  You've already got the URLs for the references listed in the code which might be good enough for people to find and read the references
* It's not really clear to me which is better, what do you think?
----------
* At some point (future PR) it would be good to use the ioda Python API so that you pick up any changes in the ioda file layout automatically through ioda updates.
   * You could use the ioda Python API directly OR use the library that Cory Martin (EMC) started (ioda-converters src/lib-python/ioda_conv_engines.py).

**Comment by ericlingerfelt** [comment]
Hi @srherbener, here are my comments -> 

- I will run ctest and correct the coding norms issues. 
- I will remove the Solo dependency. 
- I will remove the documents. The hyperlinks and citations should be enough for other developers. 
- I will create a short unit test in a future PR.
- I will leverage the IODA python API. Probably in the future PR in the previous bullet point. 

Thanks for your help and code review!

**Comment by ericlingerfelt** [comment]
Hi @srherbener, I am seeing an issue when I try to build `ioda-converters` either standalone or in with `ioda-bundle`. `ecbuild` returns this at the end of the `Makefile` configuration.

```
CMake Error at /usr/local/share/ecbuild/cmake/ecbuild_add_test.cmake:438 (add_test):
  Error evaluating generator expression:

    $<TARGET_FILE_DIR:_ioda_python>

  No target "_ioda_python"
Call Stack (most recent call first):
  iodaconv/test/CMakeLists.txt:561 (ecbuild_add_test)
```

Any ideas? I was trying to see the coding norms errors for this PR. 

Thanks!

**Comment by srherbener** [comment]
> Hi @srherbener, here are my comments ->
> 
> * I will run ctest and correct the coding norms issues.
> * I will remove the Solo dependency.
> * I will remove the documents. The hyperlinks and citations should be enough for other developers.
> * I will create a short unit test in a future PR.
> * I will leverage the IODA python API. Probably in the future PR in the previous bullet point.
> 
> Thanks for your help and code review!

@ericlingerfelt thanks for taking care of these items!

**Comment by srherbener** [comment]
> Hi @srherbener, I am seeing an issue when I try to build `ioda-converters` either standalone or in with `ioda-bundle`. `ecbuild` returns this at the end of the `Makefile` configuration.
> 
> ```
> CMake Error at /usr/local/share/ecbuild/cmake/ecbuild_add_test.cmake:438 (add_test):
>   Error evaluating generator expression:
> 
>     $<TARGET_FILE_DIR:_ioda_python>
> 
>   No target "_ioda_python"
> Call Stack (most recent call first):
>   iodaconv/test/CMakeLists.txt:561 (ecbuild_add_test)
> ```
> 
> Any ideas? I was trying to see the coding norms errors for this PR.
> 
> Thanks!

It looks like ioda somehow got built with BUILD_PYTHON_BINDINGS set to OFF. I think at this point, it would work to wipe out the build area and rerun ecbuild using `-DBUILD_PYTHON_BINDINGS=ON`. Then, to be sure, run `make update` followed by `make -j4` (or however many processes you want).

**Comment by ericlingerfelt** [comment]
The previous commit took care of the ioda coding norms error. I'm now working on the date issue by removing Solo. The main issue here is that GOES-16/17 data uses day of the year in its time stamp which Solo rectified previously. 

```
Orion-login-2[70] eric$ ctest -R iodaconv_goes_coding_norms
Test project /work/noaa/da/eric/repos/ioda-bundle/build
    Start 404: iodaconv_goes_coding_norms
1/1 Test #404: iodaconv_goes_coding_norms .......   Passed    0.27 sec

100% tests passed, 0 tests failed out of 1

Label Time Summary:
iodaconv    =   0.27 sec*proc (1 test)
script      =   0.27 sec*proc (1 test)

Total Test time (real) =   0.36 sec
```

**Comment by ericlingerfelt** [comment]
I have removed the dependency on `solo` and I have conducted another `ioda` coding norms test which passed. I believe that I have addressed all of the comments so far.

File: src/goes/goes.py
Comment (BenjaminTJohnson) [inline src/goes/goes.py:148]: Comment: We should probably compare this TB computation with CRTM's internal function for computing TBs, to ensure consistency.    Can you create a comment here identifying the source of this equation/relationship?

File: src/goes/goes_converter.py
Comment (BenjaminTJohnson) [inline src/goes/goes_converter.py:248]: Can you put a comment in this section of the code to identify the source of these relationships (if you know the source)?  This will help with future verification / understanding.

This comment applies to the azimuth section as well.

File: src/goes/goes_latlon.py
Comment (BenjaminTJohnson) [inline src/goes/goes_latlon.py:167]: All azimuth angle is not limited to -80 to 80, instead, it goes from 0 to 360.
```
.//src/CRTM_Parameters.f90:  REAL(fp), PUBLIC, PARAMETER :: MAX_SENSOR_AZIMUTH_ANGLE = 360.0_fp
.//src/CRTM_Parameters.f90:  REAL(fp), PUBLIC, PARAMETER :: MAX_SOURCE_AZIMUTH_ANGLE = 360.0_fp
```

File: src/goes/goes_latlon.py
Comment (BenjaminTJohnson) [inline src/goes/goes_latlon.py:158]: Unless I'm missing something, This is incorrect.  

See eq. (1) of the following reference for the conversion:
https://www.researchgate.net/publication/258662879_Comparison_of_two_transmittance_algorithms_in_the_Community_Radiative_Transfer_Model_Application_to_AVHRR

For now, we neglect the index of refraction for air, so the n/n portion is 1.   `H_i` is negligible compared to the height of the satellite, so that's zero.   We may wish to update this in the future, but for now, this will get us close enough.  

The equation would thus be (please verify):  `sensor_view_angle = np.arcsin(np.sin(sensor_zenith_angle) * Rearth/(Rearth+Hsat))`, where Rearth is the nominal radius of earth in km, and Hsat is the altitude of the satellite above the surface in km.    I believe both of these parameters are provided in the metadata for the instrument.

File: src/goes/goes_latlon.py
Comment (BenjaminTJohnson) [inline src/goes/goes_latlon.py:168]: Same azimuth angle comment.

File: src/goes/goes_util.py
Comment (BenjaminTJohnson) [inline src/goes/goes_util.py:44]: Comment: probably good to verify the usage of fill value here, and that it's consistent with code expectations if you haven't done so already.

**Review by BenjaminTJohnson — CHANGES_REQUESTED** [review]

**Comment by BenjaminTJohnson** [comment]
@ericlingerfelt I've added my review.  Please feel free to ask any questions for clarification etc., I will respond ASAP.

File: src/goes/goes.py
Comment (ericlingerfelt) [reply src/goes/goes.py:55]: Yes. You are right. The second array index is exclusive.

File: src/goes/goes.py
Comment (ericlingerfelt) [reply src/goes/goes.py:148]: The sources for all formulations used in each python class are listed in the header of each class. Yes. I would like to compare this to the TB (assuming that means Temperature Brightness) to the one used in CRTM.

File: src/goes/goes_converter.py
Comment (ericlingerfelt) [reply src/goes/goes_converter.py:248]: The sources for all formulations used in each python class are listed in the header documentation of each class.

File: src/goes/goes_latlon.py
Comment (ericlingerfelt) [reply src/goes/goes_latlon.py:167]: OK. I'll change that acceptable range.

File: src/goes/goes_latlon.py
Comment (ericlingerfelt) [reply src/goes/goes_latlon.py:168]: OK. I'll change that acceptable range.

File: src/goes/goes_util.py
Comment (ericlingerfelt) [reply src/goes/goes_util.py:44]: I will add more comments in the code regarding this.

File: src/goes/goes_latlon.py
Comment (ericlingerfelt) [reply src/goes/goes_latlon.py:158]: I could not find the correct formulation for sensor view angle and I also saw a few test data examples where they were identical. Thank you for providing the real answer for this formula! I will check out the reference and get back to you.

**Comment by ericlingerfelt** [comment]
> @ericlingerfelt I've added my review. Please feel free to ask any questions for clarification etc., I will respond ASAP.

Thanks, @BenjaminTJohnson!

File: src/goes/goes_latlon.py
Comment (ericlingerfelt) [reply src/goes/goes_latlon.py:158]: The equation you reference is for sensor zenith angle and not sensor view angle. The variable theta_sat is labelled as scan angle and not sensor zenith angle. So I am slightly confused. Here are the links I used to calculate these values. https://satepsanone.nesdis.noaa.gov/pub/GASP/documentations/GOESR/GOES-R_SemiStatic_Data_README2.docx and https://www.ncdc.noaa.gov/gridsat/docs/Angle_Calculations.pdf.

Comment (BenjaminTJohnson) [reply src/goes/goes_latlon.py:158]: Maybe I'm confused too.   Alright, let's table this for now.   I recommend creating an issue for this specific topic, because we'll need to ensure that this part is done correctly.

**Review by BenjaminTJohnson — APPROVED** [review]

File: src/goes/goes_latlon.py
Comment (ericlingerfelt) [reply src/goes/goes_latlon.py:158]: I created this issue to verify these formulations. 

https://github.com/JCSDA-internal/ioda-converters/issues/625

File: src/goes/test_goes_converter.py
Comment (gthompsnJCSDA) [inline src/goes/test_goes_converter.py:39]: It appears that you have an extra assignment statement here and I can only assume that you need only one of these two - not sure which one.

File: src/goes/goes.py
Comment (gthompsnJCSDA) [inline src/goes/goes.py:132]: Are we really planning to handle incoming mesoscale region data?  The lat/lon values would need to be computed nearly daily for real-time data as the NWS picks regions of weather to suit the needs of operational forecasting.  Since I know that we are creating static lat/lon files intended to be read from disk to save time, then I hope the naming of the lat/lon static files has some identifier for full disk and conus at minimum since these should remain very nearly static always.

File: src/goes/goes_converter.py
Comment (gthompsnJCSDA) [inline src/goes/goes_converter.py:93]: The comment here is mistaken.  Channels 1-6 are reflectances whereas channels 7-16 are the infrared brightness temperatures.

File: src/goes/goes_converter.py
Comment (gthompsnJCSDA) [inline src/goes/goes_converter.py:301]: I know the chunk_dim is used for the compression algorithm, but shouldn't it be set lower than the number of total locations?  Isn't it more optimal to chunk the vector is smaller units that compress smaller by using semi-local min/max/range info?

File: src/goes/goes_converter.py
Comment (gthompsnJCSDA) [inline src/goes/goes_converter.py:458]: Is this flag meanings section done correctly?  If I was to guess, I would expect flag value 0 to mean 'good_pixel', flag value 1 to mean 'conditionally_usable_pixel', flag value 2 to mean 'out_of_range_pixel' and flag value 3 to mean 'no_value_pixel.'  The way it's written (here plus a duplicate in next method) appears to have a space between the out-of-range and no-value so I would be expecting the list to be 4 items for flag_meanings.

**Review by gthompsnJCSDA — COMMENTED** [review]
I sprinkled in comments where I think attention may be needed.

It would be tremendously helpful to visualize what the code does by adding some plots.  For example, if you start with the 2km common remapped (all 16 channels) and make a plot of one or two of them, and then make another plot with the re-sampled from 2km to 4km, then up to 8km.  No need to go super big, but just some visualizations to see the code is doing as advertised.  Similarly, it might be good to see plots of the final output of some of the angles to ensure that all appears to work.

**Comment by srherbener** [comment]
@gthompsnJCSDA are you okay with merging this PR (once the CI test failures are addressed). Unless the issues you brought up are show-stoppers, my preference would be to get this merged now and address outstanding issues in subsequent PRs. Thanks!

**Comment by ericlingerfelt** [comment]
Please wait on the merge. @fabiolrdiniz and I almost have all of the angles corrected with images. I will be committing to this branch later today. Thanks for your patience!

**Comment by srherbener** [comment]
> Please wait on the merge. @fabiolrdiniz and I almost have all of the angles corrected with images. I will be committing to this branch later today. Thanks for your patience!

Thanks for the update @ericlingerfelt. I'll wait for the commit before merging.

File: src/goes/goes_converter.py
Comment (ericlingerfelt) [reply src/goes/goes_converter.py]: This has been resolved in the latest commit.

File: src/goes/goes_converter.py
Comment (ericlingerfelt) [reply src/goes/goes_converter.py]: This has been resolved.

File: src/goes/goes.py
Comment (ericlingerfelt) [reply src/goes/goes.py:132]: This will be addressed in a future PR. Currently, we are only using full disk scans for the Data Lake project. Please remind me of this later on.

File: src/goes/test_goes_converter.py
Comment (ericlingerfelt) [reply src/goes/test_goes_converter.py]: This has been resolved.

File: src/goes/goes_converter.py
Comment (ericlingerfelt) [reply src/goes/goes_converter.py:317]: This is a placeholder until an optimized chuck dimension has been selected. This will be addressed in a future PR.

**Comment by ericlingerfelt** [comment]
@srherbener, please merge at your leisure.

**Comment by srherbener** [comment]
@ericlingerfelt we are still getting the coding norms errors:
```
Test output
+ SRC_DIR=/jcsda/ioda-bundle/iodaconv/src/goes
+ PYCODESTYLE_CFG_PATH=/jcsda/ioda-bundle/iodaconv
+ cd /jcsda/ioda-bundle/iodaconv/src/goes
+ pycodestyle -v --config=/jcsda/ioda-bundle/iodaconv/.pycodestyle --filename=*.py,*.py.in .
cli configuration: /jcsda/ioda-bundle/iodaconv/.pycodestyle
directory .
checking ./goes.py
checking ./goes_converter.py
./goes_converter.py:257:29: W503 line break before binary operator
./goes_converter.py:258:29: W503 line break before binary operator
./goes_converter.py:264:19: W503 line break before binary operator
./goes_converter.py:265:19: W503 line break before binary operator
./goes_converter.py:268:23: W503 line break before binary operator
checking ./goes_latlon.py
checking ./goes_util.py
checking ./test_goes_converter.py
5       W503 line break before binary operator
```

These are a pain, but there appears to be two ways to fix this (https://www.python.org/dev/peps/pep-0008/#maximum-line-length)
* use line continuation (backslash at the end of the line)
* use parantheses (ie, around the entire expression on the lhs of the assignment)

Here's the first one the checker is complaining about:
https://github.com/JCSDA-internal/ioda-converters/blob/f03cd4e5be0524bed03eac4a125c79228a7c2bae/src/goes/goes_converter.py#L256-L258

So you should be able to fix it by placing another set of `()` around the entire expression.

**Comment by ericlingerfelt** [comment]
Hi @srherbener, I'll take care of this right now.

**Comment by ericlingerfelt** [comment]
I don't think the extra parenthesis will do it. I have them around the expression at line 264.

**Comment by srherbener** [comment]
How about the line continuation (backslash at the end of the line)?

**Comment by ericlingerfelt** [comment]
Hi @srherbener, the latest commit removes the issue by breaking up the equations into components.

**Comment by srherbener** [comment]
@ericlingerfelt thank you for your patience with the coding norms errors. That seems like a pain to have to break up the computations like you did. 

I'll wait for the tests to finish and merge if all looks okay.

**Pull request merged**
